### ISSUE TITLE
[RabbitMQ] TCP Socket churn when executing health check

### ISSUE THREAD

**Issue by rwkarg** [Xabaril/AspNetCore.Diagnostics.HealthChecks#445]
commit https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/commit/2a66bc9588701f7a50d33eb2652727a562ce5f33 reverted the reuse of a single connection when checking health to address the shutdown issue reported in #412 

While the shutdown issue is fixed (which was caused by the IConnectionFactory being passed in requesting Foreground threads for the RabbitMQ socket management), it does now result in TCP socket churn since a new connection is created for every health check evaluation. Eventually this results in ephemeral port exhaustion (given frequent enough health checks) and the process becomes unable to initiate TCP connections.

**Comment by rwkarg** [comment]
Looking at potential resolutions to reintroduce Connection reuse without also reintroducing the undesired shutdown behavior reported in #412 

There doesn't seem to be a direct way to clean up a health check on application shutdown.

Ideally the RabbitMQHealthCheck (or any implementation of IHealthCheck) could be made `IDisposable` and then the framework provided DefaultHealthCheckService would implement `IDisposable` as well and handle necessary disposal of registered health checks. Alternatively the `HealthCheckServiceOptions` could implement `IDisposable` and handle the same clean up.
Unfortunately, that would be an update to .NET and not arrive until .NET 5.0.

For now, it may be worth looking at either plumbing in a generic IDisposable handler trigger by a passed in IApplicationLifetime instance (may be too broad of a solution for this situation? unintended consequences?) or more specific handling of RabbitMQHealthCheck clean up. This would allow the connection reuse to be put back without reintroducing the undesired shutdown behavior reported in #412

**Comment by rwkarg** [comment]
I'm going to take a look at what implementing specific shutdown handling of `RabbitMQHealthCheck` instances in the `HealthCheckServiceOptions.Registrations` might look like.

**Comment by rwkarg** [comment]
One difficulty is that while I can make changes to the extension method (`.AddRabbitMQ`) to wire up the cleanup, there isn't a straight forward way to include such clean up in the example code from #412 where they add the `RabbitMQHealthCheck` outside of the extension method, similar to this:

```
services.AddHealthChecks().Add(
                new HealthCheckRegistration(
                    "rabbitmq",
                    provider => new RabbitMQHealthCheck(provider.GetService<IConnectionFactory>()),
                    HealthStatus.Healthy,
                    new string[0]));

**Comment by rwkarg** [comment]
Would it be acceptable to say that if you use the `.AddRabbitMQ` extension method then the package will take care of the `RabbitMQHealthCheck` lifetime, but if you choose to create your own `RabbitMQHealthCheck` and add it with `.Add/new HealthCheckRegistration` then you're in charge of handling the lifetime of the `RabbitMQHealthCheck`?

**Comment by rwkarg** [comment]
@ig-sinicyn Is using the provided `.AddRabbitMQ` extension possible in your use case?

**Comment by ig-sinicyn** [comment]
Hi, @rwkarg 

We've discovered the issue with port exhaustion, too. So we've made custom health check that uses 
an instance of rabbitmq connection (passed via DI).

So feel free to change the code and thanks a lot for the notification:)

**Comment by rwkarg** [comment]
I will have some time this week to work on this, though there hasn't yet been a straight forward solution for all of the various ways that this health check can be registered through the extension methods.
Tagging @unaizorrilla for visibility.

**Comment by unaizorrilla** [comment]
Let me know for help! I don't read all the discussion but, at this moment the health check allow:

 - set connection string and create transient connections
 - Use registered IConnection or IConnectionFactory on DI
 - Allow specify a Action to create a connection
 - Allow specify a Action to create a connection factory

Any of these methods can fit for the problem?

**Comment by rwkarg** [comment]
There are two ways to address this. The first way is to configure the ConnectionFactory to use background threads for IO which will not hold the process open when it's trying to shutdown. In fact, the only reason I can think of to use Foreground threads for IO is if you explicitly want to keep the process running while the Connection is alive.

It's tempting to provide guidance that if you use Foreground threads for IO on the Connection used for health checks then you're doing it wrong, but the default for RabbitMQ.Client is to use foreground threads. Saying that defaults are bad and as well not doing anything to fix it isn't the best user experience. Additionally, it is conceivable that there is a use case where Foreground threads for IO are useful and there should be a comprehensive answer as to how to handle that.

The difficulty is that there are many ways to set up a RabbitMQHealthCheck and each of them has a different "ownership" model for the underlying Connection as well as different access (or lack of access) to application lifetime events to properly clean up connections that are "owned" by this health check framework.

For example, in the connection string health check registration, I can also require the IHostApplicationLifetime be passed in so that I can call dispose on the RabbitMQHealthCheck (after implementing IDisposable there) to clean up the connection if there is one internally. But the registration that takes in and Action<IServiceProvider, IConection> or pulls the connection from DI doesn't "own" that connection so it can't clean it up. That case needs to rely on either the Action implementor or the DI container, respectively, to clean up the IConnection (or rely on the connection being configured with background IO threads which won't keep the process open).

Not looking for specific answers here. Just documenting the thought process so far.

**Comment by rwkarg** [comment]
PR up now. This will almost always do the right thing if you use the extension methods (no shutdown hang ever as well as reuse of connection if you provide either the URL or an IConnectionFactory with UseBackgroundThreadsForIO=true).

**Comment by unaizorrilla** [comment]
> 
> 
> Hi, @rwkarg
> 
> We've discovered the issue with port exhaustion, too. So we've made custom health check that uses
> an instance of rabbitmq connection (passed via DI).
> 
> So feel free to change the code and thanks a lot for the notification:)


Hi @ig-sinicyn 

RabbitMQ can work with IConnection or IConnectionFactory from DI, can you set if any of AddRabbit overloads work for your scenario?

**Comment by CarlosLanderas** [comment]
> Hi, @rwkarg
> 
> We've discovered the issue with port exhaustion, too. So we've made custom health check that uses
> an instance of rabbitmq connection (passed via DI).
> 
> So feel free to change the code and thanks a lot for the notification:)

@rwkarg looking at the PR I wonder myself why can't you configure your RabbitMQ connection using DI with the lifetime that you need and the strategies that might be required being that you have a factory method.

Maybe you can give us some more context. Thanks!

**Comment by unaizorrilla** [comment]
Hi all,

I tested this again and all is working for me (gracefully shutdown and re-use same connection and network recovery) I have shutdown rabbit mq and the connection is recovered automatically!!

 ```csharp
            services.AddSingleton<IConnectionFactory>(sp =>
            {
                return new ConnectionFactory()
                {
                    Uri = new Uri("amqp://guest:guest@localhost:5672"),
                    AutomaticRecoveryEnabled = true
                };
            });

            services.AddSingleton<IConnection>(sp =>
            {
                return sp.GetRequiredService<IConnectionFactory>().CreateConnection();
            });

            services.AddHealthChecks()
                .AddRabbitMQ(sp=>sp.GetService<IConnection>());
```

Could you try this sample and tell me if is working for you!

**Comment by rwkarg** [comment]
> Hi, @rwkarg
> 
> We've discovered the issue with port exhaustion, too. So we've made custom health check that uses
> an instance of rabbitmq connection (passed via DI).
> 
> So feel free to change the code and thanks a lot for the notification:)



> > Hi, @rwkarg
> > We've discovered the issue with port exhaustion, too. So we've made custom health check that uses
> > an instance of rabbitmq connection (passed via DI).
> > So feel free to change the code and thanks a lot for the notification:)
> 
> @rwkarg looking at the PR I wonder myself why can't you configure your RabbitMQ connection using DI with the lifetime that you need and the strategies that might be required being that you have a factory method.
> 
> Maybe you can give us some more context. Thanks!

The simple example is having more than one broker that a specific service connects to. We can't use DI to register multiple instances of IConnection or IConnectionFactory to have multiple health checks registered, one for each broker.
Even with a single broker, it should be expected that there are multiple connections. The recommended way to connect to RabbitMQ is to have one connection dedicated to publishing and one dedicated to consuming so that if flow control blocks the publishing connection, the consuming connection is not impacted (flow control slows publishers so that consumers can catch up). 

Additionally, even for our services that communicate with only a single broker we don't have either an IConnection or an IConnectionFactory that we register in DI. Our RabbitMQ framework gets a connection string at bootstrapping but there is not an exposed IConnectionFactory or IConnection to register. 

We _could_ take the URI at bootstrapping, create our own ConnectionFactory, and create our own Connection, and pass that Connection in to .AddRabbitMQ(..); but that is what the PR ends up doing for you (in the right way) when you pass in a URI to .AddRabbitMQ(..).

**Comment by rwkarg** [comment]
> Hi all,
> 
> I tested this again and all is working for me (gracefully shutdown and re-use same connection and network recovery) I have shutdown rabbit mq and the connection is recovered automatically!!
> 
> ```cs
>            services.AddSingleton<IConnectionFactory>(sp =>
>            {
>                return new ConnectionFactory()
>                {
>                    Uri = new Uri("amqp://guest:guest@localhost:5672"),
>                    AutomaticRecoveryEnabled = true
>                };
>            });
> 
>            services.AddSingleton<IConnection>(sp =>
>            {
>                return sp.GetRequiredService<IConnectionFactory>().CreateConnection();
>            });
> 
>            services.AddHealthChecks()
>                .AddRabbitMQ(sp=>sp.GetService<IConnection>());
> ```
> 
> Could you try this sample and tell me if is working for you!

Passing in a Singleton IConnection (or a specific instance that was created outside of the `.AddRabbitMQ` call) is the only scenario that reuses a connection.
Passing in an IConnectionFactory (or resolving that from DI) or a URI/connection string will result in no connection reuse and eventual port exhaustion (observed by us and @ig-sinicyn above).

If we're open to removing the non-IConnection options then we will have a greater chance at correct behavior but there is a cost for users of the health check to migrate from the removed methods of registration. Even still, there would need to be documentation on the "right way" to go from either a URI or IConnectionFactory to an IConnection that won't result in either of the two issues (the process failing to terminate or ephemeral port exhaustion) that are possible if the IConnection is provided the "wrong way" or with "wrong settings".

**Comment by rwkarg** [comment]
The problem is that this health check can make the machine the process is running on be unable to open TCP sockets (ephemeral port exhaustion).

With the 3.0.4 code, if you register a health check with the extension method that takes either a connection string or IConnectionFactory, then every time the health check is evaluated, a new Connection is created and then disposed. This uses, then disposes of TCP sockets. The closed sockets will hang around in TIME_WAIT (normal TCP behavior) for a period of time which keeps their port allocation. Given frequent enough health checks, the machine will run out of ephemeral ports and be unable to allocate TCP sockets for _anything_.

The fix is to reuse an existing connection to evaluate health whenever possible. This is also in line with the recommendations from the RabbitMQ team, as mentioned in #304 and implement in subsequent PRs. That connection reuse was reverted in #412 because the connection being reused could have created Foreground thread that would keep the process from terminating.

In summary:
3.0.3 reused connections but could possibly prevent the process from terminating
3.0.4 ensures the process can always terminate but could result in ephemeral port exhaustion

This change is to address both of the shortcomings in 3.0.3 and 3.0.4.
It is made difficult because of the large public surface area exposed by this check (multiple methods of registering through extension methods as well as RabbitMQHealthCheck being `public` which allows many more ways it could be registered without using the extension method.

These changes could be "simpler" but that would involve making breaking changes to the already released surface area and behavior which I have attempted to avoid.

**Comment by rwkarg** [comment]
With 3.0.4, this will result in a new connection being created for every health check evaluation:
```
services.AddSingleton<IConnectionFactory>(sp =>
           {
               return new ConnectionFactory()
               {
                   Uri = new Uri("amqp://guest:guest@localhost:5672"),
                   AutomaticRecoveryEnabled = true
               };
           });

           services.AddHealthChecks()
               .AddRabbitMQ(sp=>sp.GetService<IConnectionFactory>());
```
as will this
```
           services.AddHealthChecks()
               .AddRabbitMQ(rabbitMQConnectionString:"amqp://guest:guest@localhost:5672");
```

**Comment by unaizorrilla** [comment]
> 
> 
> With 3.0.4, this will result in a new connection being created for every health check evaluation:
> 
> ```
> services.AddSingleton<IConnectionFactory>(sp =>
>            {
>                return new ConnectionFactory()
>                {
>                    Uri = new Uri("amqp://guest:guest@localhost:5672"),
>                    AutomaticRecoveryEnabled = true
>                };
>            });
> 
>            services.AddHealthChecks()
>                .AddRabbitMQ(sp=>sp.GetService<IConnectionFactory>());
> ```
> 
> as will this
> 
> ```
>            services.AddHealthChecks()
>                .AddRabbitMQ(rabbitMQConnectionString:"amqp://guest:guest@localhost:5672");
> ```

Yeap, but you can do 


```csharp
services.AddSingleton<IConnectionFactory>(sp =>
           {
               return new ConnectionFactory()
               {
                   Uri = new Uri("amqp://guest:guest@localhost:5672"),
                   AutomaticRecoveryEnabled = true
               };
           });

           services.AddSingleton<IConnection>(sp =>
           {
               return sp.GetRequiredService<IConnectionFactory>().CreateConnection();
           });

           services.AddHealthChecks()
               .AddRabbitMQ(sp=>sp.GetService<IConnection>());
```
Let me some days to improve this healthcheck to work without posible issues on public surface!

**Comment by rwkarg** [comment]
Thanks for taking a look!

I now realize that I included the wrong version numbers in my previous comments. I intended to say that the current version is 3.0.5 (not 3.0.4) and the previous version was 3.0.4 (not 3.0.3).

Ultimately, I can ensure that my services get updated to do the "right thing". I'm more concerned with any other adopters of this health check that need to look at the different ways of registering a check (6 ways I can count) and pick the one of the two "right" ways while avoiding the 4 "wrong" ways. Either fixing or removing the "wrong" ways will eliminate their possibility of failure.

**Comment by unaizorrilla** [comment]
Hi just pushed a change over this and I think all scenarios is working well right now, can you check?

The idea:

 - Healthcheck is always use IConnection and this is created when healthcheck is registered using (connectionstring or uri), this is created with foreground threads IO and automatic recovery.
- You can regster one or more brokers and each one use and reuse the connection.
- If you register your own IConnection you this is reused.


When release finished a new package 3.0.6 will be available!

**Comment by unaizorrilla** [comment]
Please, if you agree with this solution, please close the issue!

**Comment by rwkarg** [comment]
The .builder methods that take a Func<IServiceProvider,..> get called every time a health check is evaluated. This means that a new RabbitMQHealthCheck instance (and a new connection for that new instance) will be created every evaluation (no connection reuse) in the following two spots:
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/master/src/HealthChecks.RabbitMQ/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs#L92
and
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/master/src/HealthChecks.RabbitMQ/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs#L146

This is why my PR creates a GUID and Lazy<RabbitMQHealthCheck> per registration so that the Func can be called once per health check evaluation but only ever have one RabbitMQHealthCheck created for each registration.

**Comment by rwkarg** [comment]
@unaizorrilla  Functional tests added for the IConnectionFactory registration path.
#471

These two tests are currently failing but will need to be verified locally since it appears that rabbitmq infrastructure is not in place for these tests to run on AppVeyor.

**Comment by unaizorrilla** [comment]
Hi rwkarg

Yeap, the overloads using sp with IConnectionFactory are not reusing connections, well, if you register the IConnectionFactory and not the IConnection sound reasonable, but for not promote bad patterns like this I go to mark this as Obsolete and remove on future versions.

I think the healthcheck at this moment have enought overloads to register the IConnection, directory, from the sp or with a uri/connectionstring!!

**Comment by rwkarg** [comment]
Removing the IConnectionFactory based methods fixes both the remaining connection reuse and the connection churn (if the user passed in a ConnectionFactory configured to use Foreground threads).
